### PR TITLE
feat: validate linked operations 

### LINKED ISSUE #1355: validate linked operations
Operations can also be linked using the `@link` operator, however, currently, they are not validated. So validate all imported queries.

### PR DESCRIPTION
**Pull request by ssddOnTop**
W.I.P.

**Issue Reference(s):**  
Fixes #1355

## Summary by CodeRabbit

- **New Features**
	- Introduced asynchronous validation and construction for server configurations and application contexts.
	- Added new methods and public visibility to various structs, enhancing functionality and accessibility.
	- Implemented endpoint schema checks before logging configuration information.
	- Expanded the `rest` module with a new `operation` module and enhanced endpoint handling capabilities.
- **Refactor**
	- Modified initialization patterns across several components to utilize asynchronous operations and consolidated argument passing.
	- Updated struct fields and method implementations for clearer, more efficient code execution.
- **Bug Fixes**
	- Fixed endpoint validation logic to ensure correct configuration and operation handling.
- **Tests**
	- Adjusted test setups to align with the new asynchronous initialization and configuration patterns.

### CODE REVIEW

File: src/cli/server/server_config.rs [hunk 1]
```diff
@@ -33,8 +33,12 @@ impl ServerConfig {
         }
         rt.add_extensions(extensions);
 
-        let server_context = Arc::new(AppContext::new(blueprint.clone(), rt, endpoints));
-        Self { app_ctx: server_context, blueprint }
+        let app_context = Arc::new(
+            AppContext::new_with_validation(blueprint.clone(), rt, endpoints.clone()).await?,
+        );
+        // endpoints.validate(app_context.as_ref()).await?;
```
Comment (tusharmath) [inline src/cli/server/server_config.rs:39]: commented code?

File: src/rest/endpoint_set.rs [hunk 2]
```diff
@@ -1,27 +1,60 @@
...
     endpoints: Vec<Endpoint>,
+    marker: std::marker::PhantomData<Status>,
+}
+
+#[derive(Default, Clone, Debug)]
+pub struct Checked;
+#[derive(Default, Clone, Debug)]
+pub struct Unchecked;
+
+pub struct EndpointSetIter<'a> {
```
Comment (tusharmath) [inline src/rest/endpoint_set.rs:24]: don't think we need this iter API anymore.

File: src/cli/server/server.rs [hunk 3]
```diff
@@ -33,10 +33,9 @@ impl Server {
     /// Starts the server in the current Runtime
     pub async fn start(self) -> Result<()> {
         let blueprint = Blueprint::try_from(&self.config_module).map_err(CLIError::from)?;
-        let server_config = Arc::new(ServerConfig::new(
-            blueprint.clone(),
-            self.config_module.extensions.endpoints,
-        ));
+        let server_config = Arc::new(
+            ServerConfig::new(blueprint.clone(), self.config_module.extensions.endpoints).await?,
```
Comment (tusharmath) [inline src/cli/server/server.rs:37]: The validation should also happen when we perform a `check`

File: src/cli/tc.rs [hunk 4]
```diff
@@ -41,6 +41,13 @@
 
             match blueprint {
                 Ok(blueprint) => {
+                    // check the endpoints' schema
+                    let _ = config_module
+                        .extensions
+                        .endpoint_set
+                        .into_checked(&blueprint, runtime)
+                        .await?;
```
[the remark this answers is not in the document]
Comment (tusharmath) [reply src/cli/tc.rs:49]: We can move this check after the log for n + 1

**Comment by tusharmath** [comment]
Can we set the file name in the trace?

```
ERROR Invalid Configuration
Caused by:
  • Unknown field "posts" on type "User". [at 17:5]
```
The `[at 17:5]` isn't very useful

**Expected**
```
ERROR Invalid Configuration
Caused by:
  • Unknown field "posts" on type "User". [at ./routes.graphql:17:5]

```

**Comment by ssddOnTop** [comment]
> Can we set the file name in the trace?
> 
> ```
> ERROR Invalid Configuration
> Caused by:
>   • Unknown field "posts" on type "User". [at 17:5]
> ```
> 
> The `[at 17:5]` isn't very useful
> 
> **Expected**
> 
> ```
> ERROR Invalid Configuration
> Caused by:
>   • Unknown field "posts" on type "User". [at ./routes.graphql:17:5]
> ```

hard to pass path to the places where we validate.. might have to carry list of the paths just for this

**Review by tusharmath — APPROVED** [review]

**Pull request merged**